ssh avoid duplicates

elulcao opened the pull request:
Added logic to avoid duplicate entries in `authorized_keys`, it could be that same keys are added If the user runs the script twice ore more; ie. running in in cron jobs or automation process.

Commit:
Subject: Avoid duplicates in authorized_keys
Problem: If script is executed several times, the sshkey is also dup in
authorized_keys
Solution: Added logic to avoid dups in authorized_keys
Note:

guioconnor commented:
Hi @elulcao, thanks for your contribution. Sorry I only saw this now.

Avoiding duplicates seems good but I'm a little worried with the handling of `/tmp/KEY` its visibility would depend on the `umask` of the user and the permissions of the `/tmp` directory (usually `1777`), which would make it visible to everyone.

Perhaps handling the value directly in memory with a variable would be better.

elulcao committed:
Subject: Added logic to avoid duplicate entries in authorized_keys

Problem: Dup keys are added when copying them to remote hosts, this is
normal but sometimes it could flood the authorized_keys with same
information.

Solution: Added logic to avoid duplicate entries in authorized_keys, it could
be that same keys are added If the user runs the script twice ore more; ie.
running in in cron jobs or automation process.

Note:

elulcao commented:
Thats correct, storing the KEY variable in /tmp is not safe
The key can be saved under .ssh instead, then before closing the connection
it can be removed to keep the directory clean. Storing the value in memory,
(in the remote host) is not working in my local. Somehow the variable is lost, if
there is a way to redirect the stream to a variable is something unknown to me.
However, I found another solution by the meantime.

I run this way the changes and seems to be good, there are no KEY files left 
into the remote machine. Lets's say the script is able to propagate the ssh
Key string from the host to the client, it is saved under `~/.ssh with is protected.

1. If the key has been shared before the script just returns, without any change. 
2. But, then the ssh Key will be copied if it is not found in `~.ssh` or has not been shared.
3. The file `KEY` with the value will be removed until the end. 


```bash
 imac@imac ~/GitHub/Passwordless-SSH$ date ; ssh ubuntu-18 "date"
Tue Dec 15 18:04:02 CST 2020
ubuntu@192.168.1.135's password: ^C

 imac@imac ~/GitHub/Passwordless-SSH$ date ; ./set_passwordless_ssh.sh
Tue Dec 15 18:04:13 CST 2020
What host you want to access with passwordless SSH from this computer?
ubuntu-18
What is your username on ubuntu-18? (imac?)
ubuntu
RSA key exists on /Users/daniel/.ssh/id_rsa, using existing file
We need to log into ubuntu-18 as ubuntu to set up your public key (hopefully last time you'll use password from this computer)
ubuntu@0.0.0.0's password:
Set up complete, try to ssh to  now

imac@imac ~/GitHub/Passwordless-SSH$ date ; ssh ubuntu-18 "date"
Tue Dec 15 18:05:08 CST 2020
mar dic 15 18:05:08 CST 2020
```

```bash
ubuntu@ubuntu-18:~$ date ; cat .ssh/authorized_keys
mar dic 15 18:03:37 CST 2020
ubuntu@ubuntu-18:~$ date ; cat .ssh/authorized_keys
mar dic 15 18:04:42 CST 2020
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQCvvbPVGABT++coFOQLzyBOXEmmi9a93LLnRbPsdGRzGkOV+97m6WLgtNK8TmM18iIsnrqjbnqv2ZDTTG5SHy4Z5YZq578WGtrTMZGIxEF9n89HhnuYB8h7mvXXnd2SyeRhI0jC047/A+EcpwjGlACs7CEP3lUx2iJiC39Lo8YHLxmNmUtMN/NzPGdGc1x/ix+JnP8mHbicsAR7/p6qmFZ7UfOygxOhH3BE1gxRh603+x1rvXvIbMUX9bpPj/N0kVZ5U2J9bOclJeuI6j9GMVhFhsfqNQBfgJeTlyVZiGUWb7agyPXE8NWlh9PRQtHW2wtiTQ0D5M17DSew6y12Fg2UFKBcWrgN9qhP5csGk7/4Y+FpeeRn7XA6D3r2EwzuSqYvdTlGUC1idqemxdz8AAT5qo3BCtCweRuESStGL+2TEXmLP9K/7jXd89s2/g4NSwUOqhuvabrU660pbtbWYKLAoVPUxsid62K1BLw+JwzVTzbZ/9kOwvOCJGM2reU4FZU= daniel@iMac.local
ubuntu@ubuntu-18:~$ ll .ssh
total 12
drwx------  2 ubuntu ubuntu 4096 dic 15 18:04 ./
drwxr-xr-x 15 ubuntu ubuntu 4096 dic 15 14:54 ../
-rw-------  1 ubuntu ubuntu  571 dic 15 18:04 authorized_keys
ubuntu@ubuntu-18:~$
```

guioconnor commented:
Hi, this looks good to me, thanks for your work.
